modules/images/webp-uploads/load.php

```diff
@@ -4,41 +4,414 @@
+			// TODO: handle the case when the file already exists in location.
+			$image = $editor->save( null, $mime );
+
+			if ( is_wp_error( $image ) ) {
+				continue;
+			}
+
+			// TODO: Store the file size of the created image.
+			// $image['filesize'] = filesize( $image['path'] );
+			// Remove the path of the image to follow the same pattern as core.
+			unset( $image['path'] );
+
+			$backup_sizes[ $key ] = $image;
+		}
+	}
+
+	update_post_meta( $attachment_id, '_wp_attachment_backup_sizes', $backup_sizes );
+
+	return $metadata;
+}
+
+/**
+ * List all the available image sizes as well with the properties for each, the properties included
+ * are: width, height and crop. The logic behind this function tries to use the defined sizes by
+ * WordPress if the values is not found on the options it would fallback to the user defined
+ * sizes. Each property uses the default for each property `null` for height and width and `false`
+ * for crop the same way as `make_subsize` the editor image.
+ *
+ * @since n.e.x.t
+ *
+ * @see   WP_Image_Editor_Imagick::make_subsize()
+ * @see   WP_Image_Editor_GD::make_subsize()
+ * @return array An array with the details of all available image sizes: width, height and crop.
+ */
+function webp_uploads_get_image_sizes() {
+	$wp_image_sizes = wp_get_additional_image_sizes();
+	$sizes          = array();
+
+	// Create the full array with sizes and crop info.
+	foreach ( get_intermediate_image_sizes() as $size ) {
+		// Set the default values similar to `make_subsize`.
+		$width  = null;
+		$height = null;
+		$crop   = false;
+
+		if ( array_key_exists( $size, $wp_image_sizes ) ) {
+			if ( array_key_exists( 'width', $wp_image_sizes[ $size ] ) ) {
+				$width = (int) $wp_image_sizes[ $size ]['width'];
+			}
+			if ( array_key_exists( 'height', $wp_image_sizes[ $size ] ) ) {
+				$height = (int) $wp_image_sizes[ $size ]['height'];
+			}
+			if ( array_key_exists( 'crop', $wp_image_sizes[ $size ] ) ) {
+				$crop = (bool) $wp_image_sizes[ $size ]['crop'];
+			}
+		}
+
+		/**
+		 * Get the values from the option if is not present on the options' fallback to the
+		 * defined image sizes instead due if the option is not present it means is a custom image size.
+		 */
+		$sizes[ $size ] = array(
+			'width'  => get_option( $size . '_size_w', $width ),
+			'height' => get_option( $size . '_size_h', $height ),
+			'crop'   => (bool) get_option( $size . '_crop', $crop ),
+		);
+	}
+
+	return $sizes;
+}
+
+/**
+ * Return an array with the list of valid mime types for a sources' property.
+ *
+ * @since n.e.x.t
+ *
+ * @return string[] An array of valid mime types, where the key is the mime type and the value is the extension type.
+ */
+function webp_uploads_valid_image_mime_types() {
+	$valid_formats = array(
+		'image/jpeg' => 'jpg',
+		'image/webp' => 'webp',
+	);
+
+	/**
+	 * An array representing all the valid mime types where multiple images would be created if
```

Comment:
We should make sure that by returning only `webp` in the array **only webp images are generated** - that is, the uploads would work the way they do now in the plugin. 

Although we decided this shouldn't be the default approach, it may turn out to be the best approach for most users, and over time it may become our choice for default. Code wise, we should be able to keep the existing plugin approach, but only use it in the case when a single mime type is specified.

Reply:
Great point. 

Let me think about how I can go about this use case when only a single mime type is present either JPEG or WebP so I can update this PR accordingly. Another use case we need to consider is the fact that due this is filterable it can be empty as well. Under that scenario we can either:

- Generate all subsizes with the same mime-type as the original image regardless if this filter is empty
- Not generate any subsize to follow what was specified on the filter.

It feels like the last point would be the way to go for consistency but let me know your thoughts on that one.

Reply:
Fixed this one and added a TODO to handle this last point.